### PR TITLE
8286850: [macos] Add support for signing user provided app image

### PR DESCRIPTION
**Pull request by sashamatveev**
- Added support for signing predefined application image.
 - Following command can be used to sign predefined application images: jpackage --type app-image --app-image Test.app --mac-sign [additional signing options]
 - Main class and if --mac-app-store was specified will be saved in .jpackage.xml. Both values are required for signing. Main class is to generate default identity and --mac-app-store to do correct signing for App Store.
 - Signing is done exactly same as when generating app bundle. Unsigned, signed or partially signed app images are supported. App image will be completely unsigned before signing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8286850](https://bugs.openjdk.java.net/browse/JDK-8286850): [macos] Add support for signing user provided app image
 * [JDK-8287190](https://bugs.openjdk.java.net/browse/JDK-8287190): [macos] Add support for signing user provided app image (**CSR**)

### Reviewers
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8987/head:pull/8987` \
`$ git checkout pull/8987`

Update a local copy of the PR: \
`$ git checkout pull/8987` \
`$ git pull https://git.openjdk.java.net/jdk pull/8987/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8987`

View PR using the GUI difftool: \
`$ git pr show -t 8987`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8987.diff">https://git.openjdk.java.net/jdk/pull/8987.diff</a>

</details>

### CODE REVIEW

File: test/jdk/tools/jpackage/macosx/SigningAppImageTwoStepsTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ */
+public class SigningAppImageTwoStepsTest {
+
+    @Test
+    @Parameter("true")
+    @Parameter("false")
+    public void test(boolean signAppImage) throws Exception {
+        SigningCheck.checkCertificates();
+
+        Path appimageOutput = TKit.workDir().resolve("appimage");
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/SigningAppImageTwoStepsTest.java:69]: This should be `TKit.createTempDirectory("appimage")` to allow subsequent invocations of the test to create a new "appimage" directory if the test work directory already contains "appimage" directory from the previous test run.

File: test/jdk/tools/jpackage/share/jdk/jpackage/tests/PredefinedAppImageErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ * @test
+ * @summary Test jpackage output for erroneous input with --type "app-image" and --app-image
+ * @library ../../../../helpers
+ * @build jdk.jpackage.test.*
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @compile PredefinedAppImageErrorTest.java
+ * @run main/othervm/timeout=360 -Xmx512m jdk.jpackage.test.Main
+ *  --jpt-run=jdk.jpackage.tests.PredefinedAppImageErrorTest
+ *  --jpt-before-run=jdk.jpackage.test.JPackageCommand.useToolProviderByDefault
+ */
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/share/jdk/jpackage/tests/PredefinedAppImageErrorTest.java:59]: There is no  need to duplicate the whole jtreg test declaration, it supports multiple @run comments:
```
/*
 * @test
 * @summary Test jpackage output for erroneous input with --type "app-image" and --app-image
 * @library ../../../../helpers
 * @build jdk.jpackage.test.*
 * @modules jdk.jpackage/jdk.jpackage.internal
 * @compile PredefinedAppImageErrorTest.java
 * 
 * @run main/othervm/timeout=360 -Xmx512m jdk.jpackage.test.Main
 *  --jpt-run=jdk.jpackage.tests.PredefinedAppImageErrorTest
 *  --jpt-before-run=jdk.jpackage.test.JPackageCommand.useToolProviderByDefault
 * 
 * @run main/othervm/timeout=360 -Xmx512m jdk.jpackage.test.Main
 *  --jpt-run=jdk.jpackage.tests.PredefinedAppImageErrorTest
 *  --jpt-before-run=jdk.jpackage.test.JPackageCommand.useExecutableByDefault
 */
```

File: test/jdk/tools/jpackage/macosx/SigningAppImageTwoStepsTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        JPackageCommand cmd = JPackageCommand.helloAppImage()
+                .addArguments("--app-image", appImageCmd.outputBundle().toAbsolutePath())
+                .addArguments("--mac-sign")
+                .addArguments("--mac-signing-key-user-name", SigningBase.DEV_NAME)
+                .addArguments("--mac-signing-keychain", SigningBase.KEYCHAIN)
+                .removeArgumentWithValue("--input")
+                .removeArgumentWithValue("--dest")
+                .removeArgumentWithValue("--name")
+                .removeArgumentWithValue("--main-jar")
+                .removeArgumentWithValue("--main-class")
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/SigningAppImageTwoStepsTest.java:102]: There is no point in calling `JPackageCommand.helloAppImage()` if later all arguments added by this call get deleted. Simply call JPackageCommand ctor instead.

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

**Comment by sashamatveev** [comment]
8286850: [macos] Add support for signing user provided app image [v2]
 - Fixed all comments from Alexey.
8286850: [macos] Add support for signing user provided app image [v3]
 - Fixed whitespace issues.

**Review by alexeysemenyukoracle — APPROVED** [review]

**Pull request closed without merging**